### ISSUE TITLE
Support configuring default salt

### ISSUE THREAD

**Issue by brandon-leapyear** [haskell-unordered-containers/hashable#218]
Blog post: https://cs-syd.eu/posts/2021-09-11-json-vulnerability

The biggest problem exposed by this blog post is that any third-party code using `hash` (using the default salt) that has problems with hash collisions can have malicious input constructed beforehand. The `-frandom-init-seed` flag helps with this at the cost of a different seed every start up.

One quick fix that can help mitigate this is to allow reading an environment variable like `HASHABLE_DEFAULT_SALT` with `unsafePerformIO` that one can set differently for their application. This has the advantage that my application can hash things with a different salt than the default, publicly available one, so an attacker attacking my application can't reuse malicious values from attacking another application. It would also be difficult to construct malicious input if my application's default salt isn't shared, and my application doesn't easily expose the default salt in any way.

Related:
* https://github.com/haskell-unordered-containers/unordered-containers/issues/319
* https://github.com/haskell/aeson/issues/864

**Comment by NorfairKing** [comment]
> One quick fix that can help mitigate this is to allow reading an environment variable like HASHABLE_DEFAULT_SALT with unsafePerformIO that one can set differently for their application. 

This is _SO_ hacky that we shouldn't even merge the PR, just have people use the open PR if necessary. IMO

**Comment by brandon-leapyear** [comment]
I disagree that it's "_SO_" hacky. Using environment variables to configure upstream behavior is not uncommon ([`SYSTEM_CERTIFICATE_PATH` for `x509-system`](https://github.com/vincenthz/hs-certificate/blob/7041244d0dd3946d96af66611b2760d2478e055f/x509-system/System/X509/Unix.hs#L37) or [`http_proxy` for `http-client`](https://github.com/snoyberg/http-client/blob/106951809b3e1c796ffd6cccefc4c1fad16df7c9/http-client/README.md#proxy-environment-variable)). `unsafePerformIO` isn't terribly hacky if used correctly, in this case to do one-time IO setup actions.

**Comment by jappeace** [comment]
wouldn't setting HASHABLE_RANDOM_SEED work?

**Comment by NorfairKing** [comment]
@jappeace yes that's a (very short-term) workaround.

**Comment by brandon-leapyear** [comment]
> It doesn't give the ability for configuring it, but I think you don't want to configure it anyway, just let it be random on boot.

Yes, I mentioned that in the beginning

> The -frandom-init-seed flag helps with this at the cost of a different seed every start up.

But in some situations, you don't want it to be random on boot, such as when you need deterministic tests.
